File: mandelbrot/mandel.fs
// Learn more about F# at http://fsharp.net
let (winW,winH) = (System.Console.WindowWidth, System.Console.WindowHeight)

type ViewState = { ulX : float; ulY : float; scaleX : float; scaleY : float }
let initState = { ulX = -2.00 ; ulY = 1.00; scaleX = 0.04; scaleY = 0.1 }

let procCmd key st =
  match key with
    | System.ConsoleKey.LeftArrow ->  { st with ulX = st.ulX - 6.0*st.scaleX }
    | System.ConsoleKey.RightArrow -> { st with ulX = st.ulX + 6.0*st.scaleX }
    | System.ConsoleKey.UpArrow ->    { st with ulY = st.ulY - 4.0*st.scaleY }
    | System.ConsoleKey.DownArrow ->  { st with ulY = st.ulY + 4.0*st.scaleY }
    | System.ConsoleKey.I -> { ulX = st.ulX + st.scaleX*((float)winW/4.0) ; 
                               ulY = st.ulY + st.scaleY*((float)winH/4.0) ;
                               scaleX = st.scaleX / 2.0 ;
                               scaleY = st.scaleY / 2.0 }
    | System.ConsoleKey.O -> let sx = st.scaleX * 2.0
                             let sy = st.scaleY * 2.0
                             { ulX = st.ulX - sx*((float)winW/4.0) ; 
                               ulY = st.ulY - sy*((float)winH/4.0) ;
                               scaleX = sx ; scaleY = sy }
    | System.ConsoleKey.Q -> exit 0 
    | _ -> st

let mandel xp yp = 
    let norm x y = x*x + y*y
    let rec iterate ch x y =
        if ch = ' ' 
        then ch
        else if (norm x y) > 4.0 
             then ch
             else iterate (char (int ch - 1)) (x*x - y*y + xp) (x*y*2.0 + yp)
    iterate '~' xp yp
     
let scene st = 
    let line = Array.create winW ' '
    for curY = 0 to winH-1 do
      System.Console.SetCursorPosition(0, curY)
      for curX = 0 to winW-1 do
        line.[curX]  <- mandel (st.ulX+(float curX)*st.scaleX) (st.ulY+(float curY)*st.scaleY)
      printf "%s" (new System.String(line))

let rec iteration st = 
  scene st
  let ch = System.Console.ReadKey true
  iteration ( procCmd ch.Key st )

[<EntryPoint>]
let main argv = 
    System.Console.Clear()
    iteration initState

File: mandelbrot/mandel.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Numerics;

namespace AsciiMandelbrot
{
    class Program
    {
        private char[] buffer;
        private double ulX;
        private double ulY;
        private double scaleX;
        private double scaleY;

        private char iterate(Complex pt)
        {
            Complex cur = pt;
            char ans = '~';
            while (ans > ' ')
            {
                if ((cur.Real * cur.Real + cur.Imaginary * cur.Imaginary) > 4.0) break;
                cur = cur * cur + pt;
                --ans;
            }
            return ans;
        }

        private void scene() {
            Console.SetCursorPosition(0,0);
            var wh = Console.WindowHeight - 1;
            for(int idy = 0; idy < wh; ++idy) {
                var curY = ulY + idy * scaleY;
                for (int idx = 0; idx < buffer.Length; ++idx)
                {
                    buffer[idx] = iterate(new Complex(ulX + idx * scaleX, curY));
                }
                Console.Write(buffer);
            }
            Console.Out.Flush();
        }

        public Program()
        {
            buffer = new char[Console.WindowWidth];
            ulX = -2.0; ulY = -1.0;
            scaleX = 0.04; scaleY = 0.1;
        }

        public void MainLoop()
        {
            var wh = Console.WindowHeight - 1;
            while (true) 
            {
                scene();
                var k = Console.ReadKey(true);
                switch(k.Key) {
                    case ConsoleKey.UpArrow:
                        ulY -= 3.0 * scaleY;   break;
                    case ConsoleKey.DownArrow:
                        ulY += 3.0 * scaleY;   break;
                    case ConsoleKey.LeftArrow:
                        ulX -= 6.0 * scaleX;   break;
                    case ConsoleKey.RightArrow:
                        ulX += 6.0 * scaleX;   break;
                    case ConsoleKey.I:
                        ulX += scaleX * buffer.Length / 4;
                        ulY += scaleY * wh / 4;
                        scaleX /= 2;
                        scaleY /= 2;
                        break;
                    case ConsoleKey.O:
                        scaleX *= 2;
                        scaleY *= 2;
                        ulX -= scaleX * buffer.Length / 4;
                        ulY -= scaleY * wh / 4;
                        break;
                    case ConsoleKey.Q:
                        return;
                }
            }

        }

        static void Main(string[] args)
        {
            try
            {
                Console.Clear();
                Console.CursorVisible = false;
                var pgm = new Program();
                pgm.MainLoop();
            }
            finally
            {
                Console.Clear();
                Console.CursorVisible = true;
            }
        }
    }
}
